compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs

```diff
@@ -0,0 +1,182 @@
+-- Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE TemplateHaskell #-}
+
+-- | Constraint generator for DAML LF static verification
+module DA.Daml.LF.Verify.Generate
+  ( genPackages
+  , genChoice -- Added export to suppress the unused warning.
+  ) where
+
+import Control.Lens hiding (Context)
+import Control.Monad.Error.Class (MonadError (..))
+import Data.HashMap.Strict (singleton)
+import Control.Monad.Reader
+import qualified Data.NameMap as NM
+
+import DA.Daml.LF.Ast
+import DA.Daml.LF.Verify.Context
+import DA.Daml.LF.Verify.Subst
+import DA.Daml.LF.TypeChecker.Error
+
+data GenOutput = GenOutput
+  { _goExp :: Expr
+    -- ^ The expression, evaluated as far as possible.
+  , _goUpd :: UpdateSet
+    -- ^ The updates, performed by this expression.
+  , _goDel :: Delta
+    -- ^ The context extension, made by this expression.
+  }
+
+makeLenses ''GenOutput
+
+-- | Extend a generator output with the updates and the context extensions of a
+-- second generator output. Note that the end result will contain the first
+-- expression.
+combineGO :: GenOutput -> GenOutput
+          -- ^ The two generator outputs to combine.
+          -> GenOutput
+combineGO genOut1 genOut2
+  = extendGOUpds (_goUpd genOut2)
+  $ extendGODelta (_goDel genOut2)
+    genOut1
+
+updateGOExpr :: Expr
+             -- ^ The new output expression.
+             -> GenOutput
+             -- ^ The current generator output.
+             -> GenOutput
+updateGOExpr expr = set goExp expr
+
+extendGOUpds :: UpdateSet
+             -- ^ The extension of the update set.
+             -> GenOutput
+             -- ^ The current generator output.
+             -> GenOutput
+extendGOUpds upds = over goUpd (concatUpdateSet upds)
+
+extendGODelta :: Delta
+              -- ^ The additional context extension.
+              -> GenOutput
+              -- ^ The current generator output.
+              -> GenOutput
+extendGODelta delta = over goDel (concatDelta delta)
+
+-- | Builds up the Delta environment by recursively extending the environment
+-- for each computation, and combining the output environments.
+buildDelta :: MonadDelta m => Delta -> (a -> m Delta) -> [a] -> m Delta
+buildDelta del op args = foldM step del args
+  where
+    step d x = concatDelta d <$> setDelta d (op x)
```

Comment:
My _suspicion_, fuelled by your comment on `setDelta`, is that you want something like this:
```haskell
introDeltas :: MonadDelta m => (a -> m Delta) -> [a] -> m b -> m b
introDeltas _ [] act = act
introDeltas f (x:xs) act = do
  delta <- f x
  introDelta delta (introDeltas f xs act)
```
and then have functions like
```haskell
introModule :: MonadDelta m => PackageRef -> Module -> m b -> m b
```
Is that correct?